File: client/typescript-client/src/api/orderApi.ts
import { Configuration } from "../configuration";
import { RestClient, Response } from "../rest";
import { EXCHANGE_SEGMENT, ORDER_SOURCE, ORDER_TYPE, PRODUCT } from "../settings";
import { PlaceOrderRequest, ModifyOrderRequest } from "../types";
import {
  validatePlaceOrder,
  validateCancelOrder,
  validateOrderHistory,
} from "../validation";

export class OrderApi {
  constructor(private rest: RestClient, private config: Configuration) {}

  private authHeaders(contentType: string): Record<string, string> {
    return {
      Sid: this.config.editSid || "",
      Auth: this.config.editToken || "",
      "Content-Type": contentType,
    };
  }

  private queryWithServerId(): Record<string, string> {
    return { sId: this.config.serverId || "" };
  }

  async placeOrder(r: PlaceOrderRequest): Promise<Response> {
    validatePlaceOrder(r);
    const body = {
      am: r.amo ?? "NO",
      dq: r.disclosed_quantity ?? "0",
      es: EXCHANGE_SEGMENT[r.exchange_segment],
      mp: r.market_protection ?? "0",
      pc: PRODUCT[r.product],
      pf: r.pf ?? "N",
      pr: r.price,
      pt: ORDER_TYPE[r.order_type],
      qt: r.quantity,
      rt: r.validity,
      tp: r.trigger_price ?? "0",
      ts: r.trading_symbol,
      tt: r.transaction_type,
      ig: r.tag,
      tk: r.scrip_token,
      sot: r.square_off_type,
      slt: r.stop_loss_type,
      slv: r.stop_loss_value,
      sov: r.square_off_value,
      lat: r.last_traded_price,
      tlt: r.trailing_stop_loss,
      tsv: r.trailing_sl_value,
      os: ORDER_SOURCE,
    };
    const res = await this.rest.request({
      method: "POST",
      url: this.config.getUrl("place_order"),
      queryParams: this.queryWithServerId(),
      headers: this.authHeaders("application/x-www-form-urlencoded"),
      body,
    });
    return res.data;
  }

  async cancelOrder(orderId: string, amo = "NO", verify = false, bookFetch?: () => Promise<Response>): Promise<Response> {
    return this.cancelEndpoint("cancel_order", orderId, amo, verify, bookFetch);
  }

  async cancelCoverOrder(orderId: string, amo = "NO", verify = false, bookFetch?: () => Promise<Response>): Promise<Response> {
    return this.cancelEndpoint("cancel_cover_order", orderId, amo, verify, bookFetch);
  }

  async cancelBracketOrder(orderId: string, amo = "NO", verify = false, bookFetch?: () => Promise<Response>): Promise<Response> {
    return this.cancelEndpoint("cancel_bracket_order", orderId, amo, verify, bookFetch);
  }

  private async cancelEndpoint(
    endpoint: string,
    orderId: string,
    amo: string,
    verify: boolean,
    bookFetch?: () => Promise<Response>,
  ): Promise<Response> {
    validateCancelOrder(orderId);
    if (verify && bookFetch) {
      const book = await bookFetch();
      if (Array.isArray(book?.data)) {
        for (const row of book.data as any[]) {
          if (row.nOrdNo === orderId) {
            if (["rejected", "cancelled", "complete", "traded"].includes(row.ordSt)) {
              const st = row.ordSt === "complete" ? "Traded" : row.ordSt;
              return { Error: `The Given Order Status is ${st}`, Reason: row.rejRsn };
            }
          }
        }
      }
    }
    const res = await this.rest.request({
      method: "POST",
      url: this.config.getUrl(endpoint),
      queryParams: this.queryWithServerId(),
      headers: this.authHeaders("application/x-www-form-urlencoded"),
      body: { on: orderId, am: amo },
    });
    return res.data;
  }

  async modifyOrder(r: ModifyOrderRequest, bookFetch: () => Promise<Response>): Promise<Response> {
    if (!r.order_id) throw new Error("order_id is mandatory");
    const body: Record<string, any> = {
      tk: r.instrument_token,
      mp: r.market_protection ?? "0",
      pc: r.product,
      dd: r.dd ?? "NA",
      dq: r.disclosed_quantity ?? "0",
      vd: r.validity,
      ts: r.trading_symbol,
      tt: r.transaction_type,
      pr: r.price,
      pt: r.order_type,
      fq: r.filled_quantity ?? "0",
      tp: r.trigger_price ?? "0",
      qt: r.quantity,
      no: r.order_id,
      es: r.exchange_segment,
      am: r.amo ?? "NO",
      os: ORDER_SOURCE,
    };

    const hasAll = !!(r.instrument_token && r.exchange_segment && r.trading_symbol && r.product);
    if (hasAll) {
      body.es = EXCHANGE_SEGMENT[r.exchange_segment!];
      body.pc = PRODUCT[r.product!];
      body.pt = ORDER_TYPE[r.order_type];
    } else {
      // hydrate from order book
      const book = await bookFetch();
      if (!Array.isArray(book?.data)) return { Message: "There is no Data in the Order Book" };
      const row = (book.data as any[]).find((it) => it.nOrdNo === r.order_id);
      if (!row) return { Message: `The Given Order Number ${r.order_id} is not matching with any of the orders` };
      if (["rejected", "cancelled", "complete", "traded"].includes(row.ordSt)) {
        const st = row.ordSt === "complete" ? "Traded" : row.ordSt;
        return {
          Error: `The Given Order Status is ${st}, So we can't proceed further`,
          Reason: row.rejRsn,
        };
      }
      body.ts = r.trading_symbol ?? row.trdSym;
      body.tk = r.instrument_token ?? row.tok;
      body.pc = r.product ?? row.prod;
      body.tt = r.transaction_type ?? row.trnsTp;
      body.es = r.exchange_segment ?? row.exSeg;
      if ((r.trigger_price ?? "0") === "0") body.tp = row.trgPrc;
    }

    const res = await this.rest.request({
      method: "POST",
      url: this.config.getUrl("modify_order"),
      queryParams: this.queryWithServerId(),
      headers: this.authHeaders("application/x-www-form-urlencoded"),
      body,
    });
    return res.data;
  }

  async orderReport(): Promise<Response> {
    const res = await this.rest.request({
      method: "GET",
      url: this.config.getUrl("order_book"),
      queryParams: this.queryWithServerId(),
      headers: { ...this.authHeaders("application/x-www-form-urlencoded"), accept: "application/json" },
    });
    return res.data;
  }

  async orderHistory(orderId: string): Promise<Response> {
    validateOrderHistory(orderId);
    const res = await this.rest.request({
      method: "POST",
      url: this.config.getUrl("order_history"),
      queryParams: this.queryWithServerId(),
      headers: this.authHeaders("application/x-www-form-urlencoded"),
      body: { nOrdNo: orderId },
    });
    return res.data;
  }

  async tradeReport(orderId?: string): Promise<Response> {
    const res = await this.rest.request({
      method: "GET",
      url: this.config.getUrl("trade_report"),
      queryParams: this.queryWithServerId(),
      headers: { ...this.authHeaders("application/x-www-form-urlencoded"), accept: "application/json" },
    });
    const data = res.data;
    if (!orderId) return data;
    if (!Array.isArray(data?.data))
      return { Error: "There is no trades available with the given order id" };
    const match = (data.data as any[]).find((it) => it.nOrdNo === orderId);
    if (!match) return { Error: "There is no trades available with the given order id" };
    return { stat: data.stat, stCode: data.stCode, data: match };
  }
}

File: client/java-client/src/main/java/com/kotak/neo/client/api/OrderApi.java
package com.kotak.neo.client.api;

import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.kotak.neo.client.Configuration;
import com.kotak.neo.client.Settings;
import com.kotak.neo.client.rest.RestClient;
import com.kotak.neo.client.validation.Validators;

import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.Map;

public class OrderApi {
    private final RestClient rest;
    private final Configuration config;

    public OrderApi(RestClient rest, Configuration config) {
        this.rest = rest;
        this.config = config;
    }

    private Map<String, String> authHeaders(String contentType) {
        Map<String, String> h = new HashMap<>();
        h.put("Sid", config.editSid == null ? "" : config.editSid);
        h.put("Auth", config.editToken == null ? "" : config.editToken);
        h.put("Content-Type", contentType);
        return h;
    }

    private Map<String, String> serverId() {
        Map<String, String> q = new HashMap<>();
        q.put("sId", config.serverId == null ? "" : config.serverId);
        return q;
    }

    public JsonElement placeOrder(PlaceOrderRequest r) {
        Validators.validatePlaceOrder(r.exchangeSegment, r.product, r.price, r.orderType,
                r.quantity, r.validity, r.tradingSymbol, r.transactionType);

        Map<String, Object> body = new LinkedHashMap<>();
        body.put("am", or(r.amo, "NO"));
        body.put("dq", or(r.disclosedQuantity, "0"));
        body.put("es", Settings.EXCHANGE_SEGMENT.get(r.exchangeSegment));
        body.put("mp", or(r.marketProtection, "0"));
        body.put("pc", Settings.PRODUCT.get(r.product));
        body.put("pf", or(r.pf, "N"));
        body.put("pr", r.price);
        body.put("pt", Settings.ORDER_TYPE.get(r.orderType));
        body.put("qt", r.quantity);
        body.put("rt", r.validity);
        body.put("tp", or(r.triggerPrice, "0"));
        body.put("ts", r.tradingSymbol);
        body.put("tt", r.transactionType);
        body.put("ig", r.tag);
        body.put("tk", r.scripToken);
        body.put("sot", r.squareOffType);
        body.put("slt", r.stopLossType);
        body.put("slv", r.stopLossValue);
        body.put("sov", r.squareOffValue);
        body.put("lat", r.lastTradedPrice);
        body.put("tlt", r.trailingStopLoss);
        body.put("tsv", r.trailingSLValue);
        body.put("os", Settings.ORDER_SOURCE);

        return rest.request("POST", config.getUrl("place_order"), serverId(),
                authHeaders("application/x-www-form-urlencoded"), body).data;
    }

    public JsonElement cancelOrder(String orderId, String amo, boolean verify) {
        return cancelEndpoint("cancel_order", orderId, amo, verify);
    }

    public JsonElement cancelCoverOrder(String orderId, String amo, boolean verify) {
        return cancelEndpoint("cancel_cover_order", orderId, amo, verify);
    }

    public JsonElement cancelBracketOrder(String orderId, String amo, boolean verify) {
        return cancelEndpoint("cancel_bracket_order", orderId, amo, verify);
    }

    private JsonElement cancelEndpoint(String endpoint, String orderId, String amo, boolean verify) {
        Validators.validateCancelOrder(orderId);
        if (verify) {
            JsonElement book = orderReport();
            if (book.isJsonObject() && book.getAsJsonObject().has("data")
                    && book.getAsJsonObject().get("data").isJsonArray()) {
                for (JsonElement el : book.getAsJsonObject().getAsJsonArray("data")) {
                    JsonObject row = el.getAsJsonObject();
                    if (row.has("nOrdNo") && row.get("nOrdNo").getAsString().equals(orderId)
                            && row.has("ordSt")) {
                        String st = row.get("ordSt").getAsString();
                        if (st.equals("rejected") || st.equals("cancelled")
                                || st.equals("complete") || st.equals("traded")) {
                            if (st.equals("complete")) st = "Traded";
                            JsonObject err = new JsonObject();
                            err.addProperty("Error", "The Given Order Status is " + st);
                            err.addProperty("Reason", row.has("rejRsn") ? row.get("rejRsn").getAsString() : "");
                            return err;
                        }
                    }
                }
            }
        }
        Map<String, Object> body = new HashMap<>();
        body.put("on", orderId);
        body.put("am", amo);
        return rest.request("POST", config.getUrl(endpoint), serverId(),
                authHeaders("application/x-www-form-urlencoded"), body).data;
    }

    public JsonElement modifyOrder(ModifyOrderRequest r) {
        if (r.orderId == null || r.orderId.isEmpty())
            throw new com.kotak.neo.client.exceptions.ApiValueError("order_id is mandatory");

        Map<String, Object> body = new LinkedHashMap<>();
        body.put("tk", r.instrumentToken);
        body.put("mp", or(r.marketProtection, "0"));
        body.put("pc", r.product);
        body.put("dd", or(r.dd, "NA"));
        body.put("dq", or(r.disclosedQuantity, "0"));
        body.put("vd", r.validity);
        body.put("ts", r.tradingSymbol);
        body.put("tt", r.transactionType);
        body.put("pr", r.price);
        body.put("pt", r.orderType);
        body.put("fq", or(r.filledQuantity, "0"));
        body.put("tp", or(r.triggerPrice, "0"));
        body.put("qt", r.quantity);
        body.put("no", r.orderId);
        body.put("es", r.exchangeSegment);
        body.put("am", or(r.amo, "NO"));
        body.put("os", Settings.ORDER_SOURCE);

        boolean hasAll = r.instrumentToken != null && r.exchangeSegment != null
                && r.tradingSymbol != null && r.product != null;

        if (hasAll) {
            body.put("es", Settings.EXCHANGE_SEGMENT.get(r.exchangeSegment));
            body.put("pc", Settings.PRODUCT.get(r.product));
            body.put("pt", Settings.ORDER_TYPE.get(r.orderType));
        } else {
            JsonElement book = orderReport();
            if (!book.isJsonObject() || !book.getAsJsonObject().has("data")
                    || !book.getAsJsonObject().get("data").isJsonArray()) {
                JsonObject m = new JsonObject();
                m.addProperty("Message", "There is no Data in the Order Book");
                return m;
            }
            JsonArray arr = book.getAsJsonObject().getAsJsonArray("data");
            JsonObject match = null;
            for (JsonElement el : arr) {
                JsonObject row = el.getAsJsonObject();
                if (row.has("nOrdNo") && row.get("nOrdNo").getAsString().equals(r.orderId)) {
                    match = row; break;
                }
            }
            if (match == null) {
                JsonObject m = new JsonObject();
                m.addProperty("Message", "The Given Order Number " + r.orderId + " is not matching with any of the orders");
                return m;
            }
            if (match.has("ordSt")) {
                String st = match.get("ordSt").getAsString();
                if (st.equals("rejected") || st.equals("cancelled")
                        || st.equals("complete") || st.equals("traded")) {
                    if (st.equals("complete")) st = "Traded";
                    JsonObject m = new JsonObject();
                    m.addProperty("Error", "The Given Order Status is " + st + ", So we can't proceed further");
                    m.addProperty("Reason", match.has("rejRsn") ? match.get("rejRsn").getAsString() : "");
                    return m;
                }
            }
            if (r.tradingSymbol == null && match.has("trdSym")) body.put("ts", match.get("trdSym").getAsString());
            if (r.instrumentToken == null && match.has("tok")) body.put("tk", match.get("tok").getAsString());
            if (r.product == null && match.has("prod")) body.put("pc", match.get("prod").getAsString());
            if (r.transactionType == null && match.has("trnsTp")) body.put("tt", match.get("trnsTp").getAsString());
            if (r.exchangeSegment == null && match.has("exSeg")) body.put("es", match.get("exSeg").getAsString());
            if ("0".equals(or(r.triggerPrice, "0")) && match.has("trgPrc"))
                body.put("tp", match.get("trgPrc").getAsString());
        }

        return rest.request("POST", config.getUrl("modify_order"), serverId(),
                authHeaders("application/x-www-form-urlencoded"), body).data;
    }

    public JsonElement orderReport() {
        Map<String, String> headers = authHeaders("application/x-www-form-urlencoded");
        headers.put("accept", "application/json");
        return rest.request("GET", config.getUrl("order_book"), serverId(), headers, null).data;
    }

    public JsonElement orderHistory(String orderId) {
        Validators.validateOrderHistory(orderId);
        Map<String, Object> body = new HashMap<>();
        body.put("nOrdNo", orderId);
        return rest.request("POST", config.getUrl("order_history"), serverId(),
                authHeaders("application/x-www-form-urlencoded"), body).data;
    }

    public JsonElement tradeReport(String orderId) {
        Map<String, String> headers = authHeaders("application/x-www-form-urlencoded");
        headers.put("accept", "application/json");
        JsonElement data = rest.request("GET", config.getUrl("trade_report"), serverId(), headers, null).data;
        if (orderId == null || orderId.isEmpty()) return data;
        if (!data.isJsonObject() || !data.getAsJsonObject().has("data")
                || !data.getAsJsonObject().get("data").isJsonArray()) {
            JsonObject err = new JsonObject();
            err.addProperty("Error", "There is no trades available with the given order id");
            return err;
        }
        for (JsonElement el : data.getAsJsonObject().getAsJsonArray("data")) {
            JsonObject row = el.getAsJsonObject();
            if (row.has("nOrdNo") && row.get("nOrdNo").getAsString().equals(orderId)) {
                JsonObject out = new JsonObject();
                if (data.getAsJsonObject().has("stat")) out.add("stat", data.getAsJsonObject().get("stat"));
                if (data.getAsJsonObject().has("stCode")) out.add("stCode", data.getAsJsonObject().get("stCode"));
                out.add("data", row);
                return out;
            }
        }
        JsonObject err = new JsonObject();
        err.addProperty("Error", "There is no trades available with the given order id");
        return err;
    }

    private static String or(String v, String d) { return v == null || v.isEmpty() ? d : v; }

    public static class PlaceOrderRequest {
        public String exchangeSegment, product, price, orderType, quantity, validity,
                tradingSymbol, transactionType;
        public String amo, disclosedQuantity, marketProtection, pf, triggerPrice,
                tag, scripToken, squareOffType, stopLossType, stopLossValue,
                squareOffValue, lastTradedPrice, trailingStopLoss, trailingSLValue;
    }

    public static class ModifyOrderRequest {
        public String orderId, price, orderType, quantity, validity;
        public String instrumentToken, exchangeSegment, product, tradingSymbol, transactionType,
                triggerPrice, dd, marketProtection, disclosedQuantity, filledQuantity, amo;
    }
}
